net/golioth/rpc.c

```diff
@@ -0,0 +1,253 @@
+			       COAP_TOKEN_MAX_LEN, coap_next_token(),
+			       COAP_METHOD_POST, coap_next_id());
+	if (err) {
+		return err;
+	}
+
+	err = coap_packet_append_uri_path_from_stringz(&packet, GOLIOTH_RPC_STATUS_PATH);
+	if (err) {
+		LOG_ERR("Unable add uri path to packet: %d", err);
+		return err;
+	}
+
+	err = coap_append_option_int(&packet,
+				     COAP_OPTION_CONTENT_FORMAT,
+				     COAP_CONTENT_FORMAT_APP_CBOR);
+	if (err) {
+		LOG_ERR("Unable add content format to packet: %d", err);
+		return err;
+	}
+
+	err = golioth_send_coap_payload(client, &packet, coap_payload, coap_payload_len);
+	if (err) {
+		LOG_ERR("Error in golioth_send_coap: %d", err);
+	}
+	return err;
+}
+
+static int on_rpc(const struct coap_packet *response,
+		  struct coap_reply *reply,
+		  const struct sockaddr *from)
+{
+	uint16_t payload_len;
+	const uint8_t *payload = coap_packet_get_payload(response, &payload_len);
+
+	LOG_HEXDUMP_DBG(payload, payload_len, "Payload");
+
+	if (payload_len == 3 && payload[1] == 'O' && payload[2] == 'K') {
+		/* Ignore "OK" response received after observing */
+		return 0;
+	}
+
+	UsefulBufC payload_buf = { payload, payload_len };
+
+	QCBORDecodeContext decode_ctx;
+	QCBORError qerr;
+	UsefulBufC id_buf;
+	UsefulBufC method_buf;
+
+	/* Decode id and method from request */
+	QCBORDecode_Init(&decode_ctx, payload_buf, QCBOR_DECODE_MODE_NORMAL);
+	QCBORDecode_EnterMap(&decode_ctx, NULL);
+	QCBORDecode_GetTextStringInMapSZ(&decode_ctx, "id", &id_buf);
+	QCBORDecode_GetTextStringInMapSZ(&decode_ctx, "method", &method_buf);
+	qerr = QCBORDecode_GetError(&decode_ctx);
+	if (qerr != QCBOR_SUCCESS) {
+		LOG_ERR("Failed to parse RPC request: %d (%s)", qerr, qcbor_err_to_str(qerr));
+		QCBORDecode_ExitMap(&decode_ctx);
+		QCBORDecode_Finish(&decode_ctx);
+		return qcbor_error_to_posix(qerr);
+	}
+
+	/* Start encoding response */
+	uint8_t response_buf[GOLIOTH_RPC_MAX_RESPONSE_LEN];
+	UsefulBuf response_bufc = { response_buf, sizeof(response_buf) };
+	QCBOREncodeContext encode_ctx;
+
+	QCBOREncode_Init(&encode_ctx, response_bufc);
+	QCBOREncode_OpenMap(&encode_ctx);
+	QCBOREncode_AddTextToMap(&encode_ctx, "id", id_buf);
+
+	/* Search for matching RPC registration */
+	const struct golioth_rpc *matching_rpc = NULL;
+
+	for (int i = 0; i < _num_rpcs; i++) {
+		const struct golioth_rpc *rpc = &_rpcs[i];
+
+		if (strncmp(rpc->method, method_buf.ptr, method_buf.len) == 0) {
+			matching_rpc = rpc;
+			break;
+		}
+	}
+
+	enum golioth_rpc_status status_code = GOLIOTH_RPC_UNKNOWN;
+
+	if (matching_rpc) {
+		LOG_DBG("Calling registered RPC method: %s", matching_rpc->method);
+
+		/**
+		 * Call callback while decode context is inside the params array
+		 * and encode context is inside the detail map.
+		 */
+		QCBORDecode_EnterArrayFromMapSZ(&decode_ctx, "params");
+		QCBOREncode_OpenMapInMap(&encode_ctx, "detail");
+		status_code = matching_rpc->callback(&decode_ctx,
+						     &encode_ctx,
+						     matching_rpc->callback_arg);
```

Comment:
This means that method will be called synchronously. Not sure about first version (maybe it is enough for now), but in the future we definitely want possiblity to send a reply after some long processing. This is what Azure IoT Hub allow and my experience D-Bus (on Linux as a RPC service) is that finally it is a desired thing to have. Just pointing it out for now.

Reply:
Yeah, I was thinking for that case, it would be up to the user to schedule the long-running work on the work queue or another task.

Alvaro and I ran into this issue when implementing the `reboot` method, because we were trying to reboot in the callback, which resulted in the RPC response never being sent, so we ended up scheduling the reboot in a separate task. That was a bit of an exception, because that method would effectively never complete in a synchronous way. In that case we ended replying with something along the lines of "reboot initiated", but not waiting for the reboot to complete.

We could completely remove `enum golioth_rpc_status` and expect the user to call a function like `golioth_rpc_send_response` either in the callback, or later on whenever the method is finished. This could cover both the synchronous method execution (response sent in callback) and asynchronous method execution cases (response sent after callback, possibly in another task). The main downside is that the QCBORDecodeContext and QCBOREncodeContext will go out of scope once the callback returns, so the user would need to copy stuff out and manage that themselves.

I think for now, I'll leave the synchronous approach, but I'll add a new enumerator to `enum golioth_rpc_status` for `GOLIOTH_RPC_RESPONSE_DEFERRED` and another function for `golioth_rpc_send_response`. This will at least provide some kind of option for users that need deferred execution, while keeping the synchronous case simple.

Reply:
If it's okay with you, I'm going to open a separate JIRA issue to add support for async RPC. 

I played around with a few ideas (none of which felt right) and realized it will be a substantial amount of code change.